src/main/java/fi/aalto/cs/apluscourses/presentation/ProgressViewModel.java

```diff
@@ -28,43 +29,64 @@ public ProgressViewModel() {
    */
   public Progress start(int maxValue, String label, boolean indeterminate) {
     var progress = new Progress(maxValue, label, indeterminate);
-    progresses.add(progress);
+    synchronized (lock) {
+      progresses.add(progress);
+    }
     this.updateValues();
     return progress;
   }
 
+  /**
+   * Stops a Progress.
+   */
   public void stop(Progress progress) {
-    progresses.remove(progress);
+    synchronized (lock) {
+      progresses.remove(progress);
+    }
     this.updateValues();
   }
 
+  /**
+   * Stops all Progresses.
+   */
   public void stopAll() {
-    progresses.clear();
+    synchronized (lock) {
+      progresses.clear();
+    }
     this.updateValues();
   }
 
   private void updateValues() {
-    if (progresses.isEmpty()) {
-      this.maxValue.set(0);
-      this.value.set(0);
-      this.indeterminate.set(false);
-    } else {
-      var progress = progresses.getLast();
-      this.value.set(progress.getValue());
-      this.maxValue.set(progress.getMaxValue());
-      this.label.set(progress.getLabel());
-      this.indeterminate.set(progress.getIndeterminate());
+    synchronized (lock) {
+      if (progresses.isEmpty()) {
+        this.maxValue.set(0);
+        this.value.set(0);
+        this.indeterminate.set(false);
+      } else {
+        var progress = progresses.getLast();
+        this.value.set(progress.getValue());
+        this.maxValue.set(progress.getMaxValue());
+        this.label.set(progress.getLabel());
+        this.indeterminate.set(progress.getIndeterminate());
+      }
     }
     this.updateVisible();
   }
 
   private void updateValue() {
-    var progress = progresses.getLast();
-    this.value.set(progress.getValue());
+    synchronized (lock) {
+      var progress = progresses.getLast();
+      this.value.set(progress.getValue());
+    }
   }
 
+  /**
+   * Increments a Progress and updates the value of the visible progress.
+   */
   public void increment(Progress progress) {
-    progress.increment();
+    synchronized (lock) {
+      progress.increment();
+    }
     this.updateValue();
```

Comment:
`updateValue` is called here and it assumes that `progresses` is non-empty (calls `getLast()`). However, how do we now here `progresses` is non-empty? This problem doesn't get away by adding synchronization to this class.

A possible "fix" is change this to call `updateValues` which handles the empty deque situation. `updateValue` can then be removed.

I, however, don't like too much that `Progress` class can be only accessed through this view model. May I suggest the following:
- remove `ProgressViewModel::increment`.
- add `updated` event to `Progress` class. Make the viewmodel listen to the event on `start`. Make `Progress::increment` call `updated.trigger()`.
- remove `stop` method from the `ProgressViewModel`. Instead add it `stop` (or perhaps `finish`) to `Progress` class. That method sets progress finished (`value = maxValue`) and triggers `updated` event.
- In `ProgressViewModel::updateValues`, when you get the last progress, you check if it's finished, and if true, remove it from the deque and unregister the event handler.
- Wherever you use progresses, use model (`Progress`) objects. Call `Progress::increment` and `Progress::finish` methods and let the viewmodel take care of updating itself via eventing we implemented. Use viewmodel only to start progresses (this could actually happen via some interface but let's not go that far for now) and presenting (i.e. pass it to the view).

How does this sound to you?

Reply:
Should ProgressViewModel::updateValues check if any progress is finished in that case

Reply:
Because else some finished progresses might stay in the deque

Reply:
I don't think it matters. They get eventually popped once all the progresses on top of them have been popped too.